diff --git a/importlib_metadata/__init__.py b/importlib_metadata/__init__.py
index 8761307..9a5d8d2 100644
--- a/importlib_metadata/__init__.py
+++ b/importlib_metadata/__init__.py
@@ -579,8 +579,12 @@ class Distribution:
         if context and kwargs:
             raise ValueError("cannot accept context and kwargs")
         context = context or DistributionFinder.Context(**kwargs)
-        return itertools.chain.from_iterable(
-            resolver(context) for resolver in cls._discover_resolvers()
+        return filter(
+            lambda v: hasattr(v, "_path"),
+            itertools.chain.from_iterable(
+                resolver(context)
+                for resolver in cls._discover_resolvers()
+             )
         )
 
     @staticmethod
diff --git a/pyproject.toml b/pyproject.toml
index 60de242..b240064 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -1,12 +1,10 @@
 [build-system]
-requires = ["setuptools>=56", "setuptools_scm[toml]>=3.4.1"]
+requires = ["setuptools>=56"]
 build-backend = "setuptools.build_meta"
 
 [tool.black]
 skip-string-normalization = true
 
-[tool.setuptools_scm]
-
 [tool.pytest-enabler.black]
 addopts = "--black"
 
diff --git a/setup.cfg b/setup.cfg
index efd0a36..062f10b 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -1,5 +1,6 @@
 [metadata]
 name = importlib_metadata
+version = 4.12.0
 author = Jason R. Coombs
 author_email = jaraco@jaraco.com
 description = Read metadata from Python packages
