diff --git a/lark/load_grammar.py b/lark/load_grammar.py
index d4f553c..8681b80 100644
--- a/lark/load_grammar.py
+++ b/lark/load_grammar.py
@@ -8,6 +8,7 @@ import pkgutil
 from ast import literal_eval
 from contextlib import suppress
 from typing import List, Tuple, Union, Callable, Dict, Optional, Sequence
+from importlib import resources
 
 from .utils import bfs, logger, classify_bool, is_id_continue, is_id_start, bfs_all_unique, small_factors
 from .lexer import Token, TerminalDef, PatternStr, PatternRE
@@ -23,7 +24,6 @@ from .tree import Tree, SlottedTree as ST
 from .visitors import Transformer, Visitor, v_args, Transformer_InPlace, Transformer_NonRecursive
 inline_args = v_args(inline=True)
 
-__path__ = os.path.dirname(__file__)
 IMPORT_PATHS = ['grammars']
 
 EXT = '.lark'
@@ -846,7 +846,9 @@ class FromPackageLoader:
         for path in to_try:
             full_path = os.path.join(path, grammar_path)
             try:
-                text: Optional[bytes] = pkgutil.get_data(self.pkg_name, full_path)
+                pkg = ".".join([self.pkg_name] + full_path.split(os.path.sep)[:-1])
+                item = full_path.split(os.path.sep)[-1]
+                text: Optional[bytes] = resources.read_binary(pkg, item)
             except IOError as e:
                 err = e
                 continue
