diff --git a/pyproject.toml b/pyproject.toml
index 7829c6e9..5d49544f 100644
--- a/pyproject.toml
+++ b/pyproject.toml
@@ -54,7 +54,7 @@ html5lib = "^1.0"
 importlib-metadata = { version = "^4.4", python = "<3.10" }
 jsonschema = "^4.10.0"
 # keyring uses calver, so version is unclamped
-keyring = ">=21.2.0"
+keyring = "~22.3.0"
 # packaging uses calver, so version is unclamped
 packaging = ">=20.4"
 pexpect = "^4.7.0"
diff --git a/src/poetry/console/application.py b/src/poetry/console/application.py
index 274cfcfc..9cd59231 100644
--- a/src/poetry/console/application.py
+++ b/src/poetry/console/application.py
@@ -83,7 +83,7 @@ COMMANDS = [
     "self install",
     "self lock",
     "self remove",
-    "self update",
+    # "self update",
     "self show",
     "self show plugins",
     # Source commands
diff --git a/src/poetry/console/commands/init.py b/src/poetry/console/commands/init.py
index e2ebf9cf..ef4c98f1 100644
--- a/src/poetry/console/commands/init.py
+++ b/src/poetry/console/commands/init.py
@@ -75,7 +75,7 @@ The <c1>init</c1> command creates a basic <comment>pyproject.toml</> file in the
         from poetry.core.vcs.git import GitConfig
 
         from poetry.layouts import layout
-        from poetry.utils.env import SystemEnv
+        from poetry.utils.env import SystemEnv, InterpreterLookup
 
         pyproject = PyProjectTOML(Path.cwd() / "pyproject.toml")
 
@@ -147,7 +147,8 @@ The <c1>init</c1> command creates a basic <comment>pyproject.toml</> file in the
 
         python = self.option("python")
         if not python:
-            current_env = SystemEnv(Path(sys.executable))
+            executable, _, _ = InterpreterLookup.find()
+            current_env = SystemEnv(Path(executable))
             default_python = "^" + ".".join(
                 str(v) for v in current_env.version_info[:2]
             )
diff --git a/src/poetry/console/commands/new.py b/src/poetry/console/commands/new.py
index cde571aa..0a25b2a0 100644
--- a/src/poetry/console/commands/new.py
+++ b/src/poetry/console/commands/new.py
@@ -32,7 +32,7 @@ class NewCommand(Command):
         from poetry.core.vcs.git import GitConfig
 
         from poetry.layouts import layout
-        from poetry.utils.env import SystemEnv
+        from poetry.utils.env import SystemEnv, InterpreterLookup
 
         if self.option("src"):
             layout_cls = layout("src")
@@ -65,7 +65,8 @@ class NewCommand(Command):
             if author_email:
                 author += f" <{author_email}>"
 
-        current_env = SystemEnv(Path(sys.executable))
+        executable, _, _ = InterpreterLookup.find()
+        current_env = SystemEnv(Path(executable))
         default_python = "^" + ".".join(str(v) for v in current_env.version_info[:2])
 
         layout_ = layout_cls(
diff --git a/src/poetry/repositories/installed_repository.py b/src/poetry/repositories/installed_repository.py
index 228df353..087d01de 100644
--- a/src/poetry/repositories/installed_repository.py
+++ b/src/poetry/repositories/installed_repository.py
@@ -20,9 +20,6 @@ if TYPE_CHECKING:
     from poetry.utils.env import Env
 
 
-_VENDORS = Path(__file__).parent.parent.joinpath("_vendor")
-
-
 try:
     FileNotFoundError
 except NameError:
@@ -281,13 +278,6 @@ class InstalledRepository(Repository):
                 if name in seen:
                     continue
 
-                try:
-                    path.relative_to(_VENDORS)
-                except ValueError:
-                    pass
-                else:
-                    continue
-
                 package = cls.create_package_from_distribution(distribution, env)
 
                 if with_dependencies:
diff --git a/src/poetry/utils/env.py b/src/poetry/utils/env.py
index ab645bd7..feb39ff1 100644
--- a/src/poetry/utils/env.py
+++ b/src/poetry/utils/env.py
@@ -30,6 +30,7 @@ from packaging.tags import Tag
 from packaging.tags import interpreter_name
 from packaging.tags import interpreter_version
 from packaging.tags import sys_tags
+from poetry.core.packages import Package
 from poetry.core.semver.helpers import parse_constraint
 from poetry.core.semver.version import Version
 from poetry.core.toml.file import TOMLFile
@@ -660,7 +661,10 @@ class EnvManager:
         if self._env is not None and not reload:
             return self._env
 
-        python_minor = ".".join([str(v) for v in sys.version_info[:2]])
+        python_minor = (
+            InterpreterLookup.find()[1] or
+            ".".join(str(c) for c in sys.version_info[:2])
+        )
 
         venv_path = self._poetry.config.virtualenvs_path
 
@@ -861,8 +865,7 @@ class EnvManager:
             name = self._poetry.package.name
         assert name is not None
 
-        python_patch = ".".join([str(v) for v in sys.version_info[:3]])
-        python_minor = ".".join([str(v) for v in sys.version_info[:2]])
+        python_patch, python_minor = None, None
         if executable:
             python_patch = decode(
                 subprocess.check_output(
@@ -875,7 +878,10 @@ class EnvManager:
             python_minor = ".".join(python_patch.split(".")[:2])
 
         supported_python = self._poetry.package.python_constraint
-        if not supported_python.allows(Version.parse(python_patch)):
+        if (
+            not python_patch or
+            python_patch and not supported_python.allows(Version.parse(python_patch))
+        ):
             # The currently activated or chosen Python version
             # is not compatible with the Python constraint specified
             # for the project.
@@ -887,53 +893,7 @@ class EnvManager:
                     self._poetry.package.python_versions, python_patch
                 )
 
-            io.write_error_line(
-                f"<warning>The currently activated Python version {python_patch} is not"
-                f" supported by the project ({self._poetry.package.python_versions}).\n"
-                "Trying to find and use a compatible version.</warning> "
-            )
-
-            for python_to_try in sorted(
-                self._poetry.package.AVAILABLE_PYTHONS,
-                key=lambda v: (v.startswith("3"), -len(v), v),
-                reverse=True,
-            ):
-                if len(python_to_try) == 1:
-                    if not parse_constraint(f"^{python_to_try}.0").allows_any(
-                        supported_python
-                    ):
-                        continue
-                elif not supported_python.allows_any(
-                    parse_constraint(python_to_try + ".*")
-                ):
-                    continue
-
-                python = "python" + python_to_try
-
-                if io.is_debug():
-                    io.write_line(f"<debug>Trying {python}</debug>")
-
-                try:
-                    python_patch = decode(
-                        subprocess.check_output(
-                            list_to_shell_command(
-                                [python, "-c", GET_PYTHON_VERSION_ONELINER]
-                            ),
-                            stderr=subprocess.STDOUT,
-                            shell=True,
-                        ).strip()
-                    )
-                except CalledProcessError:
-                    continue
-
-                if not python_patch:
-                    continue
-
-                if supported_python.allows(Version.parse(python_patch)):
-                    io.write_line(f"Using <c1>{python}</c1> ({python_patch})")
-                    executable = python
-                    python_minor = ".".join(python_patch.split(".")[:2])
-                    break
+            executable, python_minor, python_patch = InterpreterLookup.find(supported_python)
 
             if not executable:
                 raise NoCompatiblePythonVersionFound(
@@ -990,21 +950,13 @@ class EnvManager:
                 prompt=venv_prompt,
             )
 
-        # venv detection:
-        # stdlib venv may symlink sys.executable, so we can't use realpath.
-        # but others can symlink *to* the venv Python,
-        # so we can't just use sys.executable.
-        # So we just check every item in the symlink tree (generally <= 3)
-        p = os.path.normcase(sys.executable)
-        paths = [p]
-        while os.path.islink(p):
-            p = os.path.normcase(os.path.join(os.path.dirname(p), os.readlink(p)))
-            paths.append(p)
-
-        p_venv = os.path.normcase(str(venv))
-        if any(p.startswith(p_venv) for p in paths):
-            # Running properly in the virtualenv, don't need to do anything
-            return self.get_system_env()
+            if not root_venv:
+                envs = tomlkit.document()
+                envs_file = TOMLFile(venv_path / self.ENVS_FILE)
+                if envs_file.exists():
+                    envs = envs_file.read()
+                envs[name] = {"minor": python_minor, "patch": python_patch}
+                envs_file.write(envs)
 
         return VirtualEnv(venv)
 
@@ -1050,7 +1002,7 @@ class EnvManager:
             "--no-download",
             "--no-periodic-update",
             "--python",
-            executable or sys.executable,
+            executable or "python",
         ]
 
         if prompt is not None:
@@ -1114,9 +1066,13 @@ class EnvManager:
         want to retrieve Poetry's custom virtual environment
         (e.g. plugin installation or self update).
         """
-        prefix, base_prefix = Path(sys.prefix), Path(cls.get_base_prefix())
-        env: Env = SystemEnv(prefix)
-        if not naive:
+        pydef_executable, _, _ = InterpreterLookup.find()
+        prefix, base_prefix = (
+            Path(pydef_executable) if pydef_executable else None,
+            Path(cls.get_base_prefix())
+        )
+        env: Env = SystemEnv(prefix) if prefix else NullEnv()
+        if not naive and prefix:
             if prefix.joinpath("poetry_env").exists():
                 env = GenericEnv(base_prefix, child_env=env)
             else:
@@ -1542,97 +1498,53 @@ class SystemEnv(Env):
     A system (i.e. not a virtualenv) Python environment.
     """
 
-    @property
-    def python(self) -> str:
-        return sys.executable
+    def __init__(self, path: Path, base: Path | None = None, auto_path: bool = True) -> None:
+        self._is_windows = sys.platform == "win32"
+        if auto_path and path:
+            path = Path(
+                self._run(
+                    [str(path), "-W", "ignore", "-"],
+                    input_=GET_BASE_PREFIX
+                ).strip()
+            )
+        super().__init__(path, base=base)
 
     @property
     def sys_path(self) -> list[str]:
-        return sys.path
+        output = self.run_python_script(GET_SYS_PATH)
+        return json.loads(output)
 
     def get_version_info(self) -> tuple[Any, ...]:
-        return tuple(sys.version_info)
+        output = self.run_python_script(GET_PYTHON_VERSION)
+        return tuple([int(s) for s in output.strip().split(".")])
 
     def get_python_implementation(self) -> str:
-        return platform.python_implementation()
+        return self.marker_env["platform_python_implementation"]
 
-    def get_pip_command(self, embedded: bool = False) -> list[str]:
-        # If we're not in a venv, assume the interpreter we're running on
-        # has a pip and use that
-        return [sys.executable, self.pip_embedded if embedded else self.pip]
+    def get_marker_env(self) -> dict[str, Any]:
+        output = self.run_python_script(GET_ENVIRONMENT_INFO)
+        return json.loads(output)
 
     def get_paths(self) -> dict[str, str]:
-        # We can't use sysconfig.get_paths() because
-        # on some distributions it does not return the proper paths
-        # (those used by pip for instance). We go through distutils
-        # to get the proper ones.
-        import site
-
-        from distutils.command.install import SCHEME_KEYS
-        from distutils.core import Distribution
-
-        d = Distribution()
-        d.parse_config_files()
-        with warnings.catch_warnings():
-            warnings.filterwarnings("ignore", "setup.py install is deprecated")
-            obj = d.get_command_obj("install", create=True)
-        assert obj is not None
-        obj.finalize_options()
-
-        paths = sysconfig.get_paths().copy()
-        for key in SCHEME_KEYS:
-            if key == "headers":
-                # headers is not a path returned by sysconfig.get_paths()
-                continue
-
-            paths[key] = getattr(obj, f"install_{key}")
-
-        if site.check_enableusersite():
-            usersite = getattr(obj, "install_usersite", None)
-            userbase = getattr(obj, "install_userbase", None)
-            if usersite is not None and userbase is not None:
-                paths["usersite"] = usersite
-                paths["userbase"] = userbase
-
-        return paths
+        output = self.run_python_script(GET_PATHS)
+        return json.loads(output)
 
     def get_supported_tags(self) -> list[Tag]:
-        return list(sys_tags())
-
-    def get_marker_env(self) -> dict[str, Any]:
-        if hasattr(sys, "implementation"):
-            info = sys.implementation.version
-            iver = f"{info.major}.{info.minor}.{info.micro}"
-            kind = info.releaselevel
-            if kind != "final":
-                iver += kind[0] + str(info.serial)
-
-            implementation_name = sys.implementation.name
-        else:
-            iver = "0"
-            implementation_name = ""
-
-        return {
-            "implementation_name": implementation_name,
-            "implementation_version": iver,
-            "os_name": os.name,
-            "platform_machine": platform.machine(),
-            "platform_release": platform.release(),
-            "platform_system": platform.system(),
-            "platform_version": platform.version(),
-            "python_full_version": platform.python_version(),
-            "platform_python_implementation": platform.python_implementation(),
-            "python_version": ".".join(platform.python_version().split(".")[:2]),
-            "sys_platform": sys.platform,
-            "version_info": sys.version_info,
-            "interpreter_name": interpreter_name(),
-            "interpreter_version": interpreter_version(),
-        }
+        output = self.run_python_script(GET_SYS_TAGS)
+        return [Tag(*t) for t in json.loads(output)]
 
-    def get_pip_version(self) -> Version:
-        from pip import __version__
+    def get_pip_command(self, embedded: bool = False) -> list[str]:
+        return [
+            self._bin(self._executable),
+            self.pip_embedded if embedded else self.pip,
+        ]
 
-        return Version.parse(__version__)
+    def get_pip_version(self):
+        output = self.run_pip("--version").strip()
+        m = re.match("pip (.+?)(?: from .+)?$", output)
+        if not m:
+            return Version.parse("0.0")
+        return Version.parse(m.group(1))
 
     def is_venv(self) -> bool:
         return self._path != self._base
@@ -1853,16 +1765,100 @@ class NullEnv(SystemEnv):
         if path is None:
             path = Path(sys.prefix)
 
-        super().__init__(path, base=base)
+        super().__init__(path, base=base, auto_path=False)
 
         self._execute = execute
         self.executed: list[list[str]] = []
 
+    @property
+    def python(self) -> str:
+        return sys.executable
+
+    @property
+    def sys_path(self) -> list[str]:
+        return sys.path
+
+    def get_version_info(self) -> tuple[Any, ...]:
+        return tuple(sys.version_info)
+
+    def get_python_implementation(self) -> str:
+        return platform.python_implementation()
+
     def get_pip_command(self, embedded: bool = False) -> list[str]:
-        return [
-            self._bin(self._executable),
-            self.pip_embedded if embedded else self.pip,
-        ]
+        return [sys.executable, self.pip_embedded if embedded else self.pip]
+
+    def get_paths(self) -> dict[str, str]:
+        # We can't use sysconfig.get_paths() because
+        # on some distributions it does not return the proper paths
+        # (those used by pip for instance). We go through distutils
+        # to get the proper ones.
+        import site
+
+        from distutils.command.install import SCHEME_KEYS
+        from distutils.core import Distribution
+
+        d = Distribution()
+        d.parse_config_files()
+        with warnings.catch_warnings():
+            warnings.filterwarnings("ignore", "setup.py install is deprecated")
+            obj = d.get_command_obj("install", create=True)
+        assert obj is not None
+        obj.finalize_options()
+
+        paths = sysconfig.get_paths().copy()
+        for key in SCHEME_KEYS:
+            if key == "headers":
+                # headers is not a path returned by sysconfig.get_paths()
+                continue
+
+            paths[key] = getattr(obj, f"install_{key}")
+
+        if site.check_enableusersite():
+            usersite = getattr(obj, "install_usersite", None)
+            userbase = getattr(obj, "install_userbase", None)
+            if usersite is not None and userbase is not None:
+                paths["usersite"] = usersite
+                paths["userbase"] = userbase
+
+        return paths
+
+    def get_supported_tags(self) -> list[Tag]:
+        return list(sys_tags())
+
+    def get_marker_env(self) -> dict[str, Any]:
+        if hasattr(sys, "implementation"):
+            info = sys.implementation.version
+            iver = f"{info.major}.{info.minor}.{info.micro}"
+            kind = info.releaselevel
+            if kind != "final":
+                iver += kind[0] + str(info.serial)
+
+            implementation_name = sys.implementation.name
+        else:
+            iver = "0"
+            implementation_name = ""
+
+        return {
+            "implementation_name": implementation_name,
+            "implementation_version": iver,
+            "os_name": os.name,
+            "platform_machine": platform.machine(),
+            "platform_release": platform.release(),
+            "platform_system": platform.system(),
+            "platform_version": platform.version(),
+            "python_full_version": platform.python_version(),
+            "platform_python_implementation": platform.python_implementation(),
+            "python_version": ".".join(platform.python_version().split(".")[:2]),
+            "sys_platform": sys.platform,
+            "version_info": sys.version_info,
+            "interpreter_name": interpreter_name(),
+            "interpreter_version": interpreter_version(),
+        }
+
+    def get_pip_version(self) -> Version:
+        from pip import __version__
+
+        return Version.parse(__version__)
 
     def _run(self, cmd: list[str], **kwargs: Any) -> int | str:
         self.executed.append(cmd)
@@ -2009,3 +2005,70 @@ class MockEnv(NullEnv):
 
     def is_venv(self) -> bool:
         return self._is_venv
+
+
+class InterpreterLookup:
+    @staticmethod
+    def _version_check(executable, supported_python=None):
+        try:
+            python_patch = decode(
+                subprocess.check_output(
+                    list_to_shell_command(
+                        [
+                            executable,
+                            "-c",
+                            "\"import sys; print('.'.join([str(s) for s in sys.version_info[:3]]))\"",
+                        ]
+                    ),
+                    stderr=subprocess.STDOUT,
+                    shell=True,
+                ).strip()
+            )
+        except CalledProcessError:
+            return False, None, None
+
+        if not python_patch:
+            return False, None, None
+
+        if (
+            not supported_python or
+            (supported_python and supported_python.allows(Version.parse(python_patch)))
+        ):
+            python_minor = ".".join(python_patch.split(".")[:2])
+            return True, python_minor, python_patch
+
+        return False, None, None
+
+    @classmethod
+    def find(cls, constraint=None):
+        executable, minor, patch = None, None, None
+
+        for guess in ["python", "python3", "python2"]:
+            match, minor, patch = cls._version_check(guess, constraint)
+            if match:
+                return guess, minor, patch
+
+        for python_to_try in reversed(
+            sorted(
+                Package.AVAILABLE_PYTHONS,
+                key=lambda v: (v.startswith("3"), -len(v), v),
+            )
+        ):
+            if constraint:
+                if len(python_to_try) == 1:
+                    if not parse_constraint("^{}.0".format(python_to_try)).allows_any(
+                        constraint
+                    ):
+                        continue
+                elif not constraint.allows_all(
+                    parse_constraint(python_to_try + ".*")
+                ):
+                    continue
+
+            guess = "python" + python_to_try
+            match, minor, patch = cls._version_check(guess, constraint)
+            if match:
+                executable = guess
+                break
+
+        return executable, minor, patch
