diff --git a/.gitignore b/.gitignore
index f8aceb2..837bb1c 100644
--- a/.gitignore
+++ b/.gitignore
@@ -20,9 +20,7 @@ dist
 
 /docs/_draft.rst
 /pip-wheel-metadata
-/src/virtualenv/version.py
 /src/virtualenv/out
-venv*
 .python-version
 
 *wheel-store*
diff --git a/setup.cfg b/setup.cfg
index 1419844..c259cb5 100644
--- a/setup.cfg
+++ b/setup.cfg
@@ -1,5 +1,6 @@
 [metadata]
 name = virtualenv
+version = 20.4.7
 description = Virtual Python Environment builder
 long_description = file: README.md
 long_description_content_type = text/markdown
diff --git a/setup.py b/setup.py
index cddd1d3..1b5dbe4 100644
--- a/setup.py
+++ b/setup.py
@@ -3,10 +3,4 @@ from setuptools import __version__, setup
 if int(__version__.split(".")[0]) < 41:
     raise RuntimeError("setuptools >= 41 required to build")
 
-setup(
-    use_scm_version={
-        "write_to": "src/virtualenv/version.py",
-        "write_to_template": 'from __future__ import  unicode_literals\n\n__version__ = "{version}"\n',
-    },
-    setup_requires=["setuptools_scm >= 2"],
-)
+setup()
diff --git a/src/virtualenv/__init__.py b/src/virtualenv/__init__.py
index 5f74e3e..47b1f21 100644
--- a/src/virtualenv/__init__.py
+++ b/src/virtualenv/__init__.py
@@ -1,5 +1,16 @@
 from __future__ import absolute_import, unicode_literals
 
+import sys
+
+from pathlib import Path
+
+__path_pack__ = Path(__path__[0])
+if getattr(sys, "oxidized", False):
+    parents = 1 if sys.platform.startswith("win") else 2
+    __path_assets__ = __path_pack__.parents[parents] / "assets" / "virtualenv"
+else:
+    __path_assets__ = None
+
 from .run import cli_run, session_via_cli
 from .version import __version__
 
diff --git a/src/virtualenv/activation/via_template.py b/src/virtualenv/activation/via_template.py
index 14f0979..4b01a47 100644
--- a/src/virtualenv/activation/via_template.py
+++ b/src/virtualenv/activation/via_template.py
@@ -6,14 +6,16 @@ from abc import ABCMeta, abstractmethod
 
 from six import add_metaclass
 
+from virtualenv import __path_assets__
 from virtualenv.util.six import ensure_text
 
 from .activator import Activator
 
-if sys.version_info >= (3, 7):
-    from importlib.resources import read_binary
-else:
-    from importlib_resources import read_binary
+if not __path_assets__:
+    if sys.version_info >= (3, 7):
+        from importlib.resources import read_binary
+    else:
+        from importlib_resources import read_binary
 
 
 @add_metaclass(ABCMeta)
@@ -54,7 +56,11 @@ class ViaTemplateActivator(Activator):
 
     def instantiate_template(self, replacements, template, creator):
         # read content as binary to avoid platform specific line normalization (\n -> \r\n)
-        binary = read_binary(self.__module__, str(template))
+        if __path_assets__:
+            with (__path_assets__ / "activation" / str(template)).open("rb") as f:
+                binary = f.read()
+        else:
+            binary = read_binary(self.__module__, str(template))
         text = binary.decode("utf-8", errors="strict")
         for key, value in replacements.items():
             value = self._repr_unicode(creator, value)
diff --git a/src/virtualenv/create/__init__.py b/src/virtualenv/create/__init__.py
index 01e6d4f..9e76424 100644
--- a/src/virtualenv/create/__init__.py
+++ b/src/virtualenv/create/__init__.py
@@ -1 +1,10 @@
 from __future__ import absolute_import, unicode_literals
+
+from pathlib import Path
+
+from virtualenv import __path_assets__
+
+_PATH_ASSETS = (
+    __path_assets__ / "create" if __path_assets__ else
+    Path(__path__[0])
+)
diff --git a/src/virtualenv/create/creator.py b/src/virtualenv/create/creator.py
index 6363f8b..9b00461 100644
--- a/src/virtualenv/create/creator.py
+++ b/src/virtualenv/create/creator.py
@@ -12,6 +12,7 @@ from textwrap import dedent
 
 from six import add_metaclass
 
+from virtualenv import __path_pack__
 from virtualenv.discovery.cached_py_info import LogCmd
 from virtualenv.info import WIN_CPYTHON_2
 from virtualenv.util.path import Path, safe_delete
@@ -19,10 +20,11 @@ from virtualenv.util.six import ensure_str, ensure_text
 from virtualenv.util.subprocess import run_cmd
 from virtualenv.version import __version__
 
+from . import _PATH_ASSETS
 from .pyenv_cfg import PyEnvCfg
 
-HERE = Path(os.path.abspath(__file__)).parent
-DEBUG_SCRIPT = HERE / "debug.py"
+HERE = __path_pack__ / "create"
+DEBUG_SCRIPT = _PATH_ASSETS / "debug.py"
 
 
 class CreatorMeta(object):
diff --git a/src/virtualenv/create/via_global_ref/api.py b/src/virtualenv/create/via_global_ref/api.py
index 6f296f4..3865740 100644
--- a/src/virtualenv/create/via_global_ref/api.py
+++ b/src/virtualenv/create/via_global_ref/api.py
@@ -10,6 +10,7 @@ from virtualenv.info import fs_supports_symlink
 from virtualenv.util.path import Path
 from virtualenv.util.six import ensure_text
 
+from .. import _PATH_ASSETS
 from ..creator import Creator, CreatorMeta
 
 
@@ -100,7 +101,7 @@ class ViaGlobalRefApi(Creator):
 
     def env_patch_text(self):
         """Patch the distutils package to not be derailed by its configuration files"""
-        with self.app_data.ensure_extracted(Path(__file__).parent / "_virtualenv.py") as resolved_path:
+        with self.app_data.ensure_extracted(_PATH_ASSETS / "via_global_ref" / "_virtualenv.py") as resolved_path:
             text = resolved_path.read_text()
             return text.replace('"__SCRIPT_DIR__"', repr(os.path.relpath(str(self.script_dir), str(self.purelib))))
 
diff --git a/src/virtualenv/create/via_global_ref/builtin/python2/python2.py b/src/virtualenv/create/via_global_ref/builtin/python2/python2.py
index cacd56e..b68bc74 100644
--- a/src/virtualenv/create/via_global_ref/builtin/python2/python2.py
+++ b/src/virtualenv/create/via_global_ref/builtin/python2/python2.py
@@ -6,6 +6,7 @@ import os
 
 from six import add_metaclass
 
+from virtualenv import __path_assets__
 from virtualenv.create.describe import Python2Supports
 from virtualenv.create.via_global_ref.builtin.ref import PathRefToDest
 from virtualenv.info import IS_ZIPAPP
@@ -15,7 +16,10 @@ from virtualenv.util.zipapp import read as read_from_zipapp
 
 from ..via_global_self_do import ViaGlobalRefVirtualenvBuiltin
 
-HERE = Path(os.path.abspath(__file__)).parent
+if __path_assets__:
+    HERE = __path_assets__ / "create" / "via_global_ref"
+else:
+    HERE = Path(os.path.abspath(__file__)).parent
 
 
 @add_metaclass(abc.ABCMeta)
diff --git a/src/virtualenv/discovery/__init__.py b/src/virtualenv/discovery/__init__.py
index 01e6d4f..babf6ad 100644
--- a/src/virtualenv/discovery/__init__.py
+++ b/src/virtualenv/discovery/__init__.py
@@ -1 +1,10 @@
 from __future__ import absolute_import, unicode_literals
+
+from pathlib import Path
+
+from virtualenv import __path_assets__
+
+_PATH_ASSETS = (
+    __path_assets__ / "discovery" if __path_assets__ else
+    Path(__path__[0])
+)
diff --git a/src/virtualenv/discovery/builtin.py b/src/virtualenv/discovery/builtin.py
index 52f7398..0f1844c 100644
--- a/src/virtualenv/discovery/builtin.py
+++ b/src/virtualenv/discovery/builtin.py
@@ -97,8 +97,8 @@ def propose_interpreters(spec, try_first_with, app_data, env=None):
         if spec.is_abs:
             return
     else:
-        # 2. otherwise try with the current
-        yield PythonInfo.current_system(app_data), True
+        # 2. in binary we can't use current
+        # yield PythonInfo.current_system(app_data), True
 
         # 3. otherwise fallback to platform default logic
         if IS_WIN:
diff --git a/src/virtualenv/discovery/cached_py_info.py b/src/virtualenv/discovery/cached_py_info.py
index d16a8e2..b227b72 100644
--- a/src/virtualenv/discovery/cached_py_info.py
+++ b/src/virtualenv/discovery/cached_py_info.py
@@ -19,6 +19,8 @@ from virtualenv.util.path import Path
 from virtualenv.util.six import ensure_text
 from virtualenv.util.subprocess import Popen, subprocess
 
+from . import _PATH_ASSETS
+
 _CACHE = OrderedDict()
 _CACHE[Path(sys.executable)] = PythonInfo()
 
@@ -82,7 +84,7 @@ def _get_via_file_cache(cls, app_data, path, exe, env):
 
 
 def _run_subprocess(cls, exe, app_data, env):
-    py_info_script = Path(os.path.abspath(__file__)).parent / "py_info.py"
+    py_info_script = _PATH_ASSETS / "py_info.py"
     with app_data.ensure_extracted(py_info_script) as py_info_script:
         cmd = [exe, str(py_info_script)]
         # prevent sys.prefix from leaking into the child process - see https://bugs.python.org/issue22490
diff --git a/src/virtualenv/info.py b/src/virtualenv/info.py
index afe4097..6331d60 100644
--- a/src/virtualenv/info.py
+++ b/src/virtualenv/info.py
@@ -6,14 +6,16 @@ import platform
 import sys
 import tempfile
 
+from . import __path_pack__
+
 IMPLEMENTATION = platform.python_implementation()
 IS_PYPY = IMPLEMENTATION == "PyPy"
 IS_CPYTHON = IMPLEMENTATION == "CPython"
 PY3 = sys.version_info[0] == 3
 PY2 = sys.version_info[0] == 2
 IS_WIN = sys.platform == "win32"
-ROOT = os.path.realpath(os.path.join(os.path.abspath(__file__), os.path.pardir, os.path.pardir))
-IS_ZIPAPP = os.path.isfile(ROOT)
+ROOT = os.path.realpath(__path_pack__.parent)
+IS_ZIPAPP = False
 WIN_CPYTHON_2 = IS_CPYTHON and IS_WIN and PY2
 
 _CAN_SYMLINK = _FS_CASE_SENSITIVE = _CFG_DIR = _DATA_DIR = None
diff --git a/src/virtualenv/run/__init__.py b/src/virtualenv/run/__init__.py
index e8e7ab1..7cbcd0a 100644
--- a/src/virtualenv/run/__init__.py
+++ b/src/virtualenv/run/__init__.py
@@ -130,7 +130,7 @@ def add_version_flag(parser):
     parser.add_argument(
         "--version",
         action="version",
-        version="%(prog)s {} from {}".format(__version__, virtualenv.__file__),
+        version="%(prog)s {} from {}".format(__version__, virtualenv.__path__[0]),
         help="display the version of the virtualenv package and its location, then exit",
     )
 
diff --git a/src/virtualenv/run/plugin/activators.py b/src/virtualenv/run/plugin/activators.py
index dea2827..548e531 100644
--- a/src/virtualenv/run/plugin/activators.py
+++ b/src/virtualenv/run/plugin/activators.py
@@ -2,11 +2,30 @@ from __future__ import absolute_import, unicode_literals
 
 from argparse import ArgumentTypeError
 from collections import OrderedDict
+from importlib.metadata import EntryPoint
 
 from .base import ComponentBuilder
 
 
 class ActivationSelector(ComponentBuilder):
+    _ENTRY_POINTS = {
+        "virtualenv.activate": [
+            EntryPoint(
+                name=name,
+                value=f"virtualenv.activation.{dst}",
+                group="virtualenv.activate"
+            ) for name, dst in [
+                ("bash", "bash:BashActivator"),
+                ("cshell", "cshell:CShellActivator"),
+                ("batch", "batch:BatchActivator"),
+                ("fish", "fish:FishActivator"),
+                ("powershell", "powershell:PowerShellActivator"),
+                ("python", "python:PythonActivator"),
+                ("xonsh", "xonsh:XonshActivator")
+            ]
+        ]
+    }
+
     def __init__(self, interpreter, parser):
         self.default = None
         possible = OrderedDict(
diff --git a/src/virtualenv/run/plugin/base.py b/src/virtualenv/run/plugin/base.py
index ed10fe0..16fafaa 100644
--- a/src/virtualenv/run/plugin/base.py
+++ b/src/virtualenv/run/plugin/base.py
@@ -17,11 +17,11 @@ class PluginLoader(object):
     def entry_points_for(cls, key):
         return OrderedDict((e.name, e.load()) for e in cls.entry_points().get(key, {}))
 
-    @staticmethod
-    def entry_points():
-        if PluginLoader._ENTRY_POINTS is None:
-            PluginLoader._ENTRY_POINTS = entry_points()
-        return PluginLoader._ENTRY_POINTS
+    @classmethod
+    def entry_points(cls):
+        if cls._ENTRY_POINTS is None:
+            cls._ENTRY_POINTS = entry_points()
+        return cls._ENTRY_POINTS
 
 
 class ComponentBuilder(PluginLoader):
diff --git a/src/virtualenv/run/plugin/creators.py b/src/virtualenv/run/plugin/creators.py
index ef4177a..0a59e32 100644
--- a/src/virtualenv/run/plugin/creators.py
+++ b/src/virtualenv/run/plugin/creators.py
@@ -1,6 +1,7 @@
 from __future__ import absolute_import, unicode_literals
 
 from collections import OrderedDict, defaultdict, namedtuple
+from importlib.metadata import EntryPoint
 
 from virtualenv.create.describe import Describe
 from virtualenv.create.via_global_ref.builtin.builtin_way import VirtualenvBuiltin
@@ -11,6 +12,28 @@ CreatorInfo = namedtuple("CreatorInfo", ["key_to_class", "key_to_meta", "describ
 
 
 class CreatorSelector(ComponentBuilder):
+    _ENTRY_POINTS = {
+        "virtualenv.create": [
+            EntryPoint(
+                name=name,
+                value=f"virtualenv.create.via_global_ref.{dst}",
+                group="virtualenv.create"
+            ) for name, dst in [
+                ("venv", "venv:Venv"),
+                ("cpython3-posix", "builtin.cpython.cpython3:CPython3Posix"),
+                ("cpython3-win", "builtin.cpython.cpython3:CPython3Windows"),
+                ("cpython2-posix", "builtin.cpython.cpython2:CPython2Posix"),
+                ("cpython2-mac-framework", "builtin.cpython.mac_os:CPython2macOsFramework"),
+                ("cpython3-mac-framework", "builtin.cpython.mac_os:CPython3macOsFramework"),
+                ("cpython2-win", "builtin.cpython.cpython2:CPython2Windows"),
+                ("pypy2-posix", "builtin.pypy.pypy2:PyPy2Posix"),
+                ("pypy2-win", "builtin.pypy.pypy2:Pypy2Windows"),
+                ("pypy3-posix", "builtin.pypy.pypy3:PyPy3Posix"),
+                ("pypy3-win", "builtin.pypy.pypy3:Pypy3Windows")
+            ]
+        ]
+    }
+
     def __init__(self, interpreter, parser):
         creators, self.key_to_meta, self.describe, self.builtin_key = self.for_interpreter(interpreter)
         super(CreatorSelector, self).__init__(interpreter, parser, "creator", creators)
diff --git a/src/virtualenv/run/plugin/discovery.py b/src/virtualenv/run/plugin/discovery.py
index ac9b7f5..17c0693 100644
--- a/src/virtualenv/run/plugin/discovery.py
+++ b/src/virtualenv/run/plugin/discovery.py
@@ -1,10 +1,22 @@
 from __future__ import absolute_import, unicode_literals
 
+from importlib.metadata import EntryPoint
+
 from .base import PluginLoader
 
 
 class Discovery(PluginLoader):
-    """ """
+    _ENTRY_POINTS = {
+        "virtualenv.discovery": [
+            EntryPoint(
+                name=name,
+                value=f"virtualenv.discovery.{dst}",
+                group="virtualenv.discovery"
+            ) for name, dst in [
+                ("builtin", "builtin:Builtin")
+            ]
+        ]
+    }
 
 
 def get_discover(parser, args):
diff --git a/src/virtualenv/run/plugin/seeders.py b/src/virtualenv/run/plugin/seeders.py
index d182c6f..ee3e5bc 100644
--- a/src/virtualenv/run/plugin/seeders.py
+++ b/src/virtualenv/run/plugin/seeders.py
@@ -1,9 +1,24 @@
 from __future__ import absolute_import, unicode_literals
 
+from importlib.metadata import EntryPoint
+
 from .base import ComponentBuilder
 
 
 class SeederSelector(ComponentBuilder):
+    _ENTRY_POINTS = {
+        "virtualenv.seed": [
+            EntryPoint(
+                name=name,
+                value=f"virtualenv.seed.embed.{dst}",
+                group="virtualenv.seed"
+            ) for name, dst in [
+                ("pip", "pip_invoke:PipInvoke"),
+                ("app-data", "via_app_data.via_app_data:FromAppData")
+            ]
+        ]
+    }
+
     def __init__(self, interpreter, parser):
         possible = self.options("virtualenv.seed")
         super(SeederSelector, self).__init__(interpreter, parser, "seeder", possible)
diff --git a/src/virtualenv/seed/wheels/embed/__init__.py b/src/virtualenv/seed/wheels/embed/__init__.py
index c843c90..44fe279 100644
--- a/src/virtualenv/seed/wheels/embed/__init__.py
+++ b/src/virtualenv/seed/wheels/embed/__init__.py
@@ -1,9 +1,14 @@
 from __future__ import absolute_import, unicode_literals
 
+from virtualenv import __path_assets__
 from virtualenv.seed.wheels.util import Wheel
 from virtualenv.util.path import Path
 
-BUNDLE_FOLDER = Path(__file__).absolute().parent
+if __path_assets__:
+    BUNDLE_FOLDER = __path_assets__ / "seed" / "wheels"
+else:
+    BUNDLE_FOLDER = Path(__file__).absolute().parent
+
 BUNDLE_SUPPORT = {
     "3.10": {
         "pip": "pip-21.1.2-py3-none-any.whl",
diff --git a/tests/unit/create/test_interpreters.py b/tests/unit/create/test_interpreters.py
index 59a0991..76d40c7 100644
--- a/tests/unit/create/test_interpreters.py
+++ b/tests/unit/create/test_interpreters.py
@@ -29,6 +29,4 @@ def test_failed_to_find_implementation(of_id, mocker):
     mocker.patch("virtualenv.run.plugin.creators.CreatorSelector._OPTIONS", return_value={})
     with pytest.raises(RuntimeError) as context:
         cli_run(["-p", of_id])
-    assert repr(context.value) == repr(
-        RuntimeError("No virtualenv implementation for {}".format(PythonInfo.current_system())),
-    )
+    assert repr(context.value).startswith('RuntimeError("No virtualenv implementation for')
diff --git a/tests/unit/test_run.py b/tests/unit/test_run.py
index 53849f5..a6da734 100644
--- a/tests/unit/test_run.py
+++ b/tests/unit/test_run.py
@@ -30,9 +30,6 @@ def test_version(capsys):
     assert not extra
 
     assert __version__ in content
-    import virtualenv
-
-    assert virtualenv.__file__ in content
 
 
 @pytest.mark.parametrize("on", [True, False])
